### ISSUE TITLE
M4-4: Code Repository Constraint Analysis

### ISSUE THREAD

**Issue by bouthilx** [bouthilx/compute-forecast-framework#59]
# M4-4: Code Repository Constraint Analysis

## Objective
Analyze GitHub repositories associated with papers to find additional evidence of computational constraints not visible in papers alone.

**Worker**: Development | **Effort**: S(2-4h) | **Domain**: General

## Context
Papers often don't mention all constraints, but code repositories reveal them through configuration choices, TODOs, and implementation shortcuts. This analysis provides supplementary evidence of suppressed demand.

### From Previous Tasks
- Paper-to-repository mappings from M4-1 and M3-1
- List of papers with identified GitHub URLs
- Paper metadata for context

### External Requirements
- GitHub API access
- Repository cloning capability
- Code parsing tools

### 1. Repository Analyzer
```python
class RepositoryConstraintAnalyzer:
    def __init__(self):
        self.constraint_patterns = {
            'config_constraints': {
                'small_batch': r'batch_size["\'\s:=]+(\d+)',
                'gradient_accumulation': r'gradient_accumulation["\'\s:=]+(\d+)',
                'mixed_precision': r'fp16|mixed_precision|use_amp',
                'checkpoint_frequency': r'save_steps["\'\s:=]+(\d+)',
                'memory_optimization': r'gradient_checkpointing|cpu_offload'
            },
            'data_constraints': {
                'subsampling': r'\.sample\(|max_examples|data_subset',
                'reduced_sequence': r'max_seq_length["\'\s:=]+(\d+)',
                'smaller_images': r'image_size["\'\s:=]+(\d+)'
            },
            'training_constraints': {
                'early_stopping': r'early_stopping|patience["\'\s:=]+(\d+)',
                'fewer_epochs': r'num_epochs["\'\s:=]+(\d+)',
                'warmup_only': r'warmup_only|debug_mode'
            }
        }
    
    def analyze_repository(self, repo_path):
        constraints = {}
        for category, patterns in self.constraint_patterns.items():
            constraints[category] = self.search_patterns(repo_path, patterns)
        return constraints
```

### 2. TODO/Comment Analyzer
```python
class TodoConstraintExtractor:
    def extract_constraint_todos(self, repo_path):
        """Extract TODOs indicating compute constraints"""
        todo_patterns = [
            r'TODO.*(?:memory|GPU|compute|resources)',
            r'FIXME.*(?:batch|OOM|scale)',
            r'XXX.*(?:slow|expensive|constraint)',
            r'NOTE.*(?:limited|reduced|simplified)'
        ]
        
        todos = []
        for pattern in todo_patterns:
            matches = self.search_files(repo_path, pattern)
            todos.extend(self.extract_context(matches))
        
        return self.classify_todos(todos)
```

### 3. Configuration Comparison
```python
def compare_configs_to_standards(config_data):
    """Compare extracted configs to standard practices"""
    standards = {
        'batch_size': {
            'small': lambda x: x < 32,
            'standard': lambda x: 32 <= x <= 128,
            'large': lambda x: x > 128
        },
        'gradient_accumulation': {
            'none': lambda x: x == 1,
            'moderate': lambda x: 2 <= x <= 8,
            'heavy': lambda x: x > 8  # Compensating for small batches
        },
        'training_steps': {
            'truncated': lambda x: x < 100000,
            'standard': lambda x: 100000 <= x <= 1000000,
            'extended': lambda x: x > 1000000
        }
    }
    
    return classify_by_standards(config_data, standards)
```

### 4. Implementation Shortcut Detector
```python
class ShortcutDetector:
    def detect_shortcuts(self, code_files):
        """Identify implementation shortcuts due to constraints"""
        shortcuts = {
            'approximate_attention': [
                r'class.*(?:Linear|Local|Sparse)Attention',
                r'use_flash_attention|xformers'
            ],
            'reduced_precision': [
                r'dtype=.*float16|bfloat16',
                r'autocast|automatic_mixed_precision'
            ],
            'simplified_architecture': [
                r'# Simplified version',
                r'# Reduced from original',
                r'lightweight|efficient|small'
            ],
            'disabled_features': [
                r'if False:.*# Too expensive',
                r'return.*# Skip.*memory',
                r'pass.*# Disabled.*compute'
            ]
        }
        
        return self.search_and_classify(code_files, shortcuts)
```

### 1. Repository Constraint Evidence
```json
{
    "repo_analysis": {
        "paper_id": "mila_2024_001",
        "repo_url": "github.com/example/repo",
        "constraints_found": {
            "config_constraints": {
                "batch_size": 8,  // Very small
                "gradient_accumulation": 16,  // Compensating
                "mixed_precision": true,
                "interpretation": "Memory-constrained training"
            },
            "data_constraints": {
                "max_examples": 10000,  // Subset of full dataset
                "max_seq_length": 128,  // Reduced from standard 512
                "interpretation": "Data reduced for faster training"
            },
            "todos": [
                {
                    "file": "train.py",
                    "line": 145,
                    "text": "TODO: Increase batch size when more GPUs available",
                    "category": "resource_constraint"
                }
            ],
            "shortcuts": [
                "approximate_attention",
                "reduced_precision",
                "disabled_robustness_eval"
            ]
        },
        "constraint_score": 0.82  // High constraints
    }
}
```

### 2. Aggregate Code Patterns
```json
{
    "code_constraint_patterns": {
        "mila_repos": {
            "small_batch_prevalence": 0.73,
            "gradient_accumulation_usage": 0.68,
            "mixed_precision_usage": 0.81,
            "data_subsampling": 0.62,
            "todos_mentioning_constraints": 0.54
        },
        "benchmark_repos": {
            "small_batch_prevalence": 0.18,
            "gradient_accumulation_usage": 0.22,
            "mixed_precision_usage": 0.45,
            "data_subsampling": 0.12,
            "todos_mentioning_constraints": 0.08
        }
    }
}
```

## Quality Validation
- [ ] Test on diverse repository structures
- [ ] Validate pattern matching accuracy
- [ ] Cross-check with paper claims
- [ ] Ensure fair comparison methodology

## Dependencies
- **Depends on**: Paper-to-repo mappings from M4-1, M3-1
- **Provides to**: M4-3 (additional suppression evidence)

## Success Criteria
- Analyze 60%+ of papers with available repositories
- Find constraint evidence in 70%+ of Mila repos
- Show clear differences vs benchmark repos
- Provide interpretable constraint indicators

## Implementation Notes
- Cache repository clones for efficiency
- Handle various code structures (scripts, notebooks, configs)
- Respect rate limits for GitHub API
- Design for incremental processing

## Ethical Considerations
- Only analyze public repositories
- Don't identify individual researchers
- Focus on aggregate patterns
- Respect repository licenses
